temporian/core/data/node.py

```diff
@@ -12,62 +12,233 @@
-from temporian.core.data.feature import Feature
-from temporian.core.data.sampling import Sampling
-from temporian.core.data.sampling import IndexDType
+from temporian.core.data.dtype import DType, IndexDType
+from temporian.core.data.schema import Schema, FeatureSchema, IndexSchema
 from temporian.utils import string
 
 if TYPE_CHECKING:
-    from temporian.core.evaluation import EvaluationInput
+    from temporian.core.evaluation import EvaluationInput, EvaluationResult
     from temporian.core.operators.base import Operator
     from temporian.implementation.numpy.data.event_set import EventSet
 
 T_SCALAR = (int, float)
 
 
+@dataclass
+class Sampling:
+    """A sampling is a reference to the way data is sampled."""
+
+    creator: Optional[Operator] = None
+
+    def __hash__(self):
+        return id(self)
+
+    def __repr__(self):
+        return f"Sampling(id={id(self)}, creator={self.creator})"
+
+
+@dataclass
+class Feature:
+    """A feature is a reference to sampled data."""
+
+    creator: Optional[Operator] = None
+
+    def __hash__(self):
+        return id(self)
+
+    def __repr__(self):
+        return f"Feature(id={id(self)}, creator={self.creator})"
+
+
 class Node(object):
-    """Schema definition of an event set in the preprocessing graph.
-
-    A node represents the structure, or schema, of a collection of indexed
-    multivariate time series, or EventSets. A node does not contain any actual
-    data, but is instead used as a reference to describe the format of the
-    input, intermediate results, or output of a Graph.
-
-    Informally, a node defines the name and data types of each time series, as
-    well as the key and data type of the index (if any).
-
-    There are several ways to create a node:
-    - Through the `.node()` method in an EventSet.
-    - Through applying operators to other nodes.
-    - Manually using the `tp.input_node(...)` method to specify the name and
-        data types of each time series and the key and data type of the index.
-    - (Not recommended) By instantiating the Node class directly.
+    """A node is a reference to the input/output of ops in a compute graph.
+
+    A node does not contain any data. Use "node.evaluate(...)" to get the values
+    transiting in a node.
+
+    Use "tp.input_node" to create a node manually, or use "event_set.node()" to
+    create a node compatible with a given event set.
     """
 
     def __init__(
         self,
+        schema: Schema,
         features: List[Feature],
         sampling: Sampling,
         name: Optional[str] = None,
         creator: Optional[Operator] = None,
     ):
+        """Creates a new node manualy.
+
+        When possible, use one of the helper functions below.
+        """
+
+        self._schema = schema
         self._features = features
         self._sampling = sampling
         self._creator = creator
         self._name = name
 
+    @staticmethod
+    def create_new_features_existing_sampling(
```

Comment:
*Moved at the bottom of the file.

>> also, since we're always passing new features to it (which makes sense), maybe renaming to create_with_new_sampling and create_with_existing_sampling?

There are situations where we keep the some or all the same features (e.g. tp.glue). Currently, this is done with "create_node_with_new_reference". I don't I have a good design in mind to create a utility function for this case (yet). But I think it is important to note that create_node_new_features_existing_sampling and create_node_new_features_new_sampling are creating new features.